### PR TITLE
[doc] ShardedDDP minor additions

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Following a couple of discussions (for instance in https://github.com/pytorch/pytorch/pull/46750), adding better explanations and suggestions directly in ShardedDDP docstrings

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/optim/adascale.py [hunk 1]
```diff
@@ -554,10 +554,11 @@ def set_num_gradients_to_accumulate(self, num_gradients_to_accumulate: int, upda
         `set_scale` needs to be called to update the scale as well.
 
         TODO (min): need a way of determine how much to increase the step size?
+
         TODO (min): have both `set_scale` and `set_num_gradients_to_accumulate`
-                    is hard to use and easy to make mistake. I think it is better
-                    to specific a specify a `base_scale`. But more discussion is
-                    needed here.
+        is hard to use and easy to make mistake. I think it is better
```
Comment (blefaudeux) [inline fairscale/optim/adascale.py:559]: sphinx was complaining about that, just a warning removal

File: fairscale/optim/oss.py [hunk 2]
```diff
@@ -239,7 +239,6 @@ def clip_grad_norm(
 
         .. warning: This needs to be called on all ranks, since synchronization primitives will be used
 
-        .. warning: Model paralelism -groups other than world- are not yet supported
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:242]: not true anymore

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 3]
```diff
@@ -46,6 +45,21 @@ class ShardedDataParallel(nn.Module):
             Synchronize the models in between the ranks when starting up. Not needed if each rank has the same seed,
             or the training restarts from a saved state
 
+
+    .. warning:
+        ShardedDDP implements gradient sharding, meaning that each rank only owns a unique shard of the model gradients
+        after the backward pass, in order to save memory and some communication bandwidth.
+
+    .. warning:
+        As a consequence of sharding, in case of gradient clipping, one has to use the `clip_grad_norm` exposed by
```
Comment (anj-s) [inline fairscale/nn/data_parallel/sharded_ddp.py:54]: If users miss this docstring, will they hit an error message? How can we communicate this and the below hanging behavior to users? my fear is that they may miss these warnings and end up with a hanging run which are hard to debug.

Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:54]: it's hard to catch because this is typically done outside of OSS/ShardedDDP, what happens is that the grad scaler will check the gradients (on each rank), and will bail out if there are NaNs/inf. If it's not the case on all ranks, then some of them bail out of the step and others repeat the step, they're not in sync and the communication primitives (broadcast) hang and time out. 
I'm not sure how to improve on that, some frameworks (lightning and huggingface for instance) help with their fairscale integration because they handle all of that and the user does not have to know

**Review by anj-s — APPROVED** [review]

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 3 above)
Comment (anj-s) [reply fairscale/nn/data_parallel/sharded_ddp.py:54]: Thanks for the context! We will have to instrument timeouts on collective operations to handle hanging issues such as these. Definitely tricky to handle all the cases though since we don't have visibility.
